go/common/enclave.go

```diff
@@ -130,3 +132,48 @@ type ProducedSecretResponse struct {
 	RequesterID gethcommon.Address
 	HostAddress string
 }
+
+const (
+	bsePrefix = "blockSubmitError: "
+	l1HeadKey = "l1Head"
+)
+
+var l1HeadErrorSeparator = fmt.Sprintf(", %s=", l1HeadKey)
+
+// BlockSubmitError is used as a standard format for error response from enclave for block submission errors
+// The L1 Head hash tells the host what the enclave knows as the canonical chain head, so it can feed it the appropriate block.
+type BlockSubmitError struct {
+	L1Head  *gethcommon.Hash
+	Wrapped error
+}
+
+func (r BlockSubmitError) Error() string {
+	head := "N/A"
+	if r.L1Head != nil && (*r.L1Head != gethcommon.Hash{}) {
+		head = r.L1Head.String()
+	}
+	return fmt.Sprintf("%s%s%s%s", bsePrefix, r.Wrapped.Error(), l1HeadErrorSeparator, head)
+}
+
+func (r BlockSubmitError) Unwrap() error {
+	return r.Wrapped
+}
+
+// RehydrateBlockSubmitError is used to rebuild a BlockSubmitError type from a string (useful after its type has been lost in RPC communication)
+// the format of the error is expected to be: "BlockSubmitError: {errorMsg}, l1Head={l1HeadHash}".
+func RehydrateBlockSubmitError(err error) (*BlockSubmitError, bool) {
+	str := err.Error()
+	split := strings.Split(str, l1HeadErrorSeparator)
+	if len(split) != 2 {
+		return nil, false
+	}
+	var hash *gethcommon.Hash
+	if len(split[1]) > 0 {
+		h := gethcommon.HexToHash(split[1])
+		hash = &h
```

Comment:
Wonder if it's cleaner to declare `hash` as a non-pointer type, then take the address on line 176, below?